src/Microsoft.Azure.Batch.SoftwareEntitlement/EntitlementTokenProperties.cs

```diff
@@ -66,19 +66,26 @@ public EntitlementTokenProperties()
             IpAddresses = ImmutableHashSet<IPAddress>.Empty;
         }
 
-        public static Errorable<EntitlementTokenProperties> Build(ITokenPropertyProvider provider)
-        {
-            return Errorable.Success(new EntitlementTokenProperties())
-                .With(provider.NotBefore()).Map((e, val) => e.FromInstant(val))
-                .With(provider.NotAfter()).Map((e, val) => e.UntilInstant(val))
-                .With(provider.IssuedAt()).Map((e, val) => e.WithIssuedAt(val))
-                .With(provider.Issuer()).Map((e, val) => e.WithIssuer(val))
-                .With(provider.Audience()).Map((e, val) => e.WithAudience(val))
-                .With(provider.ApplicationIds()).Map((e, vals) => e.WithApplications(vals))
-                .With(provider.IpAddresses()).Map((e, vals) => e.WithIpAddresses(vals))
-                .With(provider.VirtualMachineId()).Map((e, val) => e.WithVirtualMachineId(val))
-                .With(provider.TokenId()).Map((e, val) => e.WithIdentifier(val));
-        }
+        public static Result<EntitlementTokenProperties, ErrorSet> Build(ITokenPropertyProvider provider) =>
+            from notBefore in provider.NotBefore()
+            join notAfter in provider.NotAfter() on true equals true
+            join issuedAt in provider.IssuedAt() on true equals true
+            join issuer in provider.Issuer() on true equals true
+            join audience in provider.Audience() on true equals true
+            join applicationIds in provider.ApplicationIds() on true equals true
+            join ipAddresses in provider.IpAddresses() on true equals true
+            join vmid in provider.VirtualMachineId() on true equals true
+            join tokenId in provider.TokenId() on true equals true
```

Comment:
Apologies, I hit "approve" on my previous review in error.

We need to tidy up this **on true equals true** pattern before merging.

Reply:
I've got two ideas for that, although I find both more messy than the `true equals true` thing.
1. Don't use LINQ query syntax for `Join`/`With`. This means either:
   a. Rewriting consuming code so that only a couple of `Result`s are joined (`With`ed) at a time.
   b. Adding back all the overloads (and possibly more) of `With`/`Map` for higher tuple/function arities.
2. Wrap `Result` objects using an extension method to change the subsequent `SelectMany` behaviour such that errors _are_ combined. Something like:
```
var joinedResults =
    from applicationIds in provider.ApplicationIds().AsJoinable()
    from ipAddresses in provider.IpAddresses().AsJoinable()
    from vmid in provider.VirtualMachineId().AsJoinable()
    ...
```
I'm really not keen on the second option because it subverts the expected behaviour of `SelectMany` and could easily be used in a way that would break at runtime, e.g.
```
var broken =
    from thing in GetThing(/* say this returns an error */).AsJoinable()
    from output in UseThing(thing).AsJoinable()
    ...
```
The problem here is that `SelectMany` allows you to use a `thing` that doesn't exist. I'd contend that the "correct" method really is "Join", even though LINQ doesn't provide ideal query syntax for criteria-less joins.

I'd also argue that this (second) approach is the most complicated and hardest to reason about.

Reply:
Meh. I'm not in love with the `on true equals true` but lets go with that as "good enough for now". We can revisit if/when we identify a cleaner approach.